docs/advanced-features/sitemaps.md

```diff
@@ -0,0 +1,198 @@
+
+This example is using [globby](https://github.com/sindresorhus/globby) to fetch our list of routes.
+
+```bash
+npm install globby --save-dev
+```
+
+Next, we can create a Node script at `scripts/generate-sitemap.mjs`.
+This file will dynamically build a sitemap based on your `/pages` directory.
+
+```js
+import { writeFileSync } from 'fs'
+import { globby } from 'globby'
+
+async function generate() {
+  const pages = await globby([
+    'pages/*.js', // Include all pages
+    '!pages/_*.js', // Exclude _app or _document
+    '!pages/api', // Exclude API Routes
+    '!pages/404.js', // Exclude 404
+  ])
+
+  const sitemap = `
+    <?xml version="1.0" encoding="UTF-8"?>
+    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+        ${pages
+          .map((page) => {
+            const path = page.replace('pages', '').replace('.js', '')
+            const route = path === '/index' ? '' : path
+
+            return `
+              <url>
+                  <loc>${`https://nextjs.org${route}`}</loc>
+              </url>
+            `
+          })
+          .join('')}
+    </urlset>
+    `
+
+  writeFileSync('public/sitemap.xml', sitemap)
+}
+
+generate()
+```
+
+> **Note:** This Node.js script is using [ES Modules](https://nodejs.org/api/esm.html), denoted by the `.mjs` extension. You could also write this script with [CommonJS](https://nodejs.org/api/modules.html#modules_modules_commonjs_modules) if you prefer.
+
+Finally, add [`postbuild` script](https://docs.npmjs.com/cli/v7/using-npm/scripts#pre--post-scripts) in your `package.json` to run this script after `next build` completes.
+Your generated sitemap file gets created at `public/sitemap.xml`, which is then served as a [static file](/docs/basic-features/static-file-serving.md) at the root of your site.
+
+```json
+{
+  "scripts": {
+    "build": "next build",
+    "postbuild": "node ./scripts/generate-sitemap.mjs",
+    "dev": "next dev",
+    "start": "next start",
+    "lint": "next lint"
+  }
+}
+```
+
+## External Content
+
+If you have externally hosted data (e.g., a CMS), you'll need to make an API request
+before you can create your sitemap. This implementation will vary depending on your data source,
+but the idea is similar. To demonstrate, I've created an example using placeholder data.
+
+First, create a new file at `pages/sitemap.xml.js`.
+
+```js
+export async function getServerSideProps({ res }) {
+  const request = await fetch(EXTERNAL_DATA_URL)
+  const posts = await request.json()
+
+  res.setHeader('Content-Type', 'text/xml')
+  res.setHeader(
+    'Cache-Control',
+    'public, s-maxage=1200, stale-while-revalidate=600'
+  )
+  res.write(createSitemap(posts))
+  res.end()
+
+  return {
+    props: {},
+  }
+}
+
+export default function Sitemap() {
```

Comment:
i think it looks a bit weird as a page component - could this also be a api route + rewrite?